### PR TITLE
[MRG+1] Add maximum streamline length property.

### PR DESCRIPTION
**Pull request by gzahl**
Hello,
I have added the option to specify the maximum streamline length. This is handy if one e.g. only needs a single (or a few) streamlines. It is especially handy if used together with "start_points".

This may be enhanced by forcing only forward (or backward) streamline integration. But that should probably be taken care of in another patch. Do you agree?

### CODE REVIEW

**Comment by tacaswell** [comment]
Can you add a test for this?

Other than that seems perfectly reasonable and I agree doing the integration direction as a second PR is a good idea.   Smaller changes are easier to review.

**Comment by gzahl** [comment]
I will try to set up a short test.

My first interface idea for the integration direction is to change maxlength optionally in a tuple:
maxlength=[0,2]  # this would be only forward integration.
maxlength=[-4,0] # only (longer) backward integration.
maxlength=2 # this would be translated to maxlength=[-2,2], which is the old default.

This would give the most flexibility. Or is it too complicated?

**Comment by tacaswell** [comment]
The negatives would be confusing.  I think a tri-state `integration_direction in {'forward', 'reverse', 'both'}` and then the maximum_length is either all forward, all back or split evenly would be simpler and get _almost_ as much power.

**Comment by gzahl** [comment]
Ok, i have added a test and copied the interface to axes and pyplot. I hope this was the right thing to do?
Tristate sounds good. Do you have an example, how it should be done in matplotlib? 
I think 'backward' is better than 'reverse', since it is the standard term when referring to integrations.

**Comment by tacaswell** [comment]
I think just strings are good enough (just validate on the way in).

You will also need to commit the test image see http://matplotlib.org/devel/testing.html#writing-an-image-comparison-test

This should be a png only test I think.

**Comment by gzahl** [comment]
Hi,
thanks for the all your help and quick replies! I now have added two more patches:
1. The baseline image for the maxlength test
2. A complete patch + test for the integration direction

Let me know, if anything has to be changed and i will try to do it tomorrow.

**Comment by gzahl** [comment]
Ok, after ironing out my rookie mistakes, i have added two more commits.
Now only a single test fails, but it does not seem to be realted to my commits.
Is there anything i can do? Thanks!

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -4465,21 +4465,25 @@ def stackplot(self, x, *args, **kwargs):
                          label_namer=None)
     def streamplot(self, x, y, u, v, density=1, linewidth=None, color=None,
                    cmap=None, norm=None, arrowsize=1, arrowstyle='-|>',
-                   minlength=0.1, transform=None, zorder=2, start_points=None):
+                   minlength=0.1, maxlength=4.0, transform=None, zorder=2,
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: All new args need to go at the end to maintain back-compatibility.  The reason is if people are (un wisely!) passing in everything as positional args.

When we can drop python2 _and_ can break API again most of these should be converted to kwarg only, but for now they need to go at the end.

File: lib/matplotlib/pyplot.py [hunk 2]
```diff
@@ -3267,8 +3267,9 @@ def step(x, y, *args, **kwargs):
 # changes will be lost
 @_autogen_docstring(Axes.streamplot)
 def streamplot(x, y, u, v, density=1, linewidth=None, color=None, cmap=None,
-               norm=None, arrowsize=1, arrowstyle='-|>', minlength=0.1,
-               transform=None, zorder=2, start_points=None, hold=None, data=None):
+               norm=None, arrowsize=1, arrowstyle='-|>', minlength=0.1, maxlength=4.0,
```
Comment (tacaswell) [inline lib/matplotlib/pyplot.py]: re-running boilerplate will take care of this.

File: lib/matplotlib/streamplot.py [hunk 3]
```diff
@@ -21,7 +21,8 @@
 
 def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
                cmap=None, norm=None, arrowsize=1, arrowstyle='-|>',
-               minlength=0.1, transform=None, zorder=2, start_points=None):
+               minlength=0.1, maxlength=4.0, transform=None, zorder=2,
```
Comment (tacaswell) [inline lib/matplotlib/streamplot.py]: following on a them, `maxlength` needs to go at the end here as well.

File: lib/matplotlib/streamplot.py [hunk 4]
```diff
@@ -91,6 +96,15 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
     line_kw = {}
     arrow_kw = dict(arrowstyle=arrowstyle, mutation_scale=10 * arrowsize)
 
+    if integration_direction not in ['both', 'forward', 'backward']:
+        errstr = "Integration direction " \
```
Comment (tacaswell) [inline lib/matplotlib/streamplot.py]: please use `()` for line continuation not `\`

**Comment by tacaswell** [comment]
Left a few minor comments that need to be addressed.  I see what you are doing putting `minlength` and `maxlength` together, but we should not break API :disappointed: .

Could you also add an entry into https://github.com/matplotlib/matplotlib/tree/master/doc/users/whats_new so that this gets properly advertised in the next release?

**Comment by gzahl** [comment]
Hi, Sorry for the delay. It seems like i don't get any Email notifications. I have hopefully addressed all the issues.
I will try to enable my notifications.

File: lib/matplotlib/streamplot.py [hunk 5]
```diff
@@ -52,9 +53,13 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
         See :class:`~matplotlib.patches.FancyArrowPatch`.
     *minlength* : float
         Minimum length of streamline in axes coordinates.
+    *maxlength* : float
+        Maximum length of streamline in axes coordinates.
```
Comment (tacaswell) [inline lib/matplotlib/streamplot.py]: can you re-arrange the docstrings too?

**Comment by gzahl** [comment]
The tests are failing after the merge with upstream. I have to track that down later.

**Comment by afvincent** [comment]
Might the test failures be related to the fact that `FancyArrowPatch` (which is used for streamplots) was modified in parallel with your PR (#6504 that superseded #6054)?

**Comment by tacaswell** [comment]
Sorry this got dropped!

Can you rebase instead of merging upstream into your branch? 

See http://matplotlib.org/devdocs/devel/gitwash/development_workflow.html?highlight=rebase

**Comment by gzahl** [comment]
Thanks for the hint about a rebase rather than a merge.
But if i now revert the merge commit, i still have the merge/merge-revert in the history, which makes the rebase rather difficult (After conflicts with my changes, i get conflicts because of the merge commit).
Is there an easier way to do it? Sorry to bother you because of this. The easiest way would to make a new branch, and cherry-pick my changes. But i can not change the branch attached to this pull request.

**Comment by WeatherGod** [comment]
If you do an interactive rebase, you can delete the commits that you don't
want in the branch and then force-push that up to github.

**Comment by gzahl** [comment]
Ok, thanks for the hint about the interactive rebase. The rebase is done, but the test (of course) still fail.
I have looked into it for a bit, but i don't think this is (mainly) because of the #6504. The starting/end points seem to be slightly different. I will look into this asap.

**Comment by afvincent** [comment]
If the diff images do not show any issue with the arrows (especially their size), then I think you are right about #6504 not being related to the test failures.

**Comment by gzahl** [comment]
Looking at the history of streamplot.py and its test file, i make the following observations:
1. The test cases have been renamed bd63121d07cee1719248432ecf3aa61b2efb204f. I will do the same with mine, to comply to the others.
2. My observation about the different starting points seem to be related to commit 71b9ae8bd8b99843d67a635dde89e858d6f7d97b. Following this a new test case has been submitted 20959b3a6e50e255a8f5d066d286941dd3681715, which uses the starting_points option, which is also used by my tests. I therefore conclude, that this fix results in the changes of my test and will resubmit the expected image (Since it otherwise seems to be fine).

**Comment by QuLogic** [comment]
Can you rebase and remove the extra copies of the test images?

Do an interactive rebase, move the update commits right below the original addition of the files, and change them to 'fixup'.

File: doc/users/whats_new/streamplot_set_maximum_length.rst [hunk 6]
```diff
@@ -0,0 +1,4 @@
+Maximum streamline length and integration direction can now be specified
+------------------------------------------------------------------------
+
+This allows to follow the vector field for a longer time and can enhance the visibilty of the flow pattern in some use cases.
```
Comment (QuLogic) [inline doc/users/whats_new/streamplot_set_maximum_length.rst]: visibilty -> visibility

Also, wrap at 80.

File: lib/matplotlib/pyplot.py [hunk 7]
```diff
@@ -2855,7 +2855,7 @@ def hexbin(x, y, C=None, gridsize=100, bins=None, xscale='linear',
 # This function was autogenerated by boilerplate.py.  Do not edit as
 # changes will be lost
 @_autogen_docstring(Axes.hist)
-def hist(x, bins=10, range=None, normed=False, weights=None, cumulative=False,
+def hist(x, bins=None, range=None, normed=False, weights=None, cumulative=False,
```
Comment (QuLogic) [inline lib/matplotlib/pyplot.py]: Unrelated change?

File: lib/matplotlib/streamplot.py [hunk 8]
```diff
@@ -58,6 +59,10 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
         In data coordinates, the same as the ``x`` and ``y`` arrays.
     *zorder* : int
         any number
+    *maxlength* : float
+        Maximum length of streamline in axes coordinates.
+    *integration_direction* : ['foward','backward','both']
```
Comment (QuLogic) [inline lib/matplotlib/streamplot.py]: foward -> forward

+ space after commas

File: lib/matplotlib/streamplot.py [hunk 9]
```diff
@@ -95,6 +100,15 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
     line_kw = {}
     arrow_kw = dict(arrowstyle=arrowstyle, mutation_scale=10 * arrowsize)
 
+    if integration_direction not in ['both', 'forward', 'backward']:
+        errstr = ("Integration direction '%s' not recognised." %
+          integration_direction)
```
Comment (QuLogic) [inline lib/matplotlib/streamplot.py]: Misaligned.

File: lib/matplotlib/streamplot.py [hunk 10]
```diff
@@ -95,6 +100,15 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
...
+        errstr += "Expected 'both', 'forward' or 'backward'."
```
Comment (QuLogic) [inline lib/matplotlib/streamplot.py]: Why not place it all in one string from the onset?

File: lib/matplotlib/streamplot.py [hunk 11]
```diff
@@ -435,17 +450,27 @@ def integrate(x0, y0):
...
+            s, xt, yt = _integrate_rk12(x0, y0, dmap, backward_time, maxlength)
+            stotal += s
+            x_traj += xt[::-1]
+            y_traj += yt[::-1]
+
+        if integration_direction in ['both', 'forward']:
+            dmap.reset_start_point(x0, y0)
+            s, xt, yt = _integrate_rk12(x0, y0, dmap, forward_time, maxlength)
+            if len(x_traj) > 0:
+              xt = xt[1:]
```
Comment (QuLogic) [inline lib/matplotlib/streamplot.py]: Indent should be a multiple of 4.

File: lib/matplotlib/tests/test_streamplot.py [hunk 12]
```diff
@@ -16,6 +16,14 @@ def velocity_field():
     V = 1 + X - Y**2
     return X, Y, U, V
 
+def swirl_velocity_field():
+    x = np.linspace(-3.,3.,100)
+    y = np.linspace(-3.,3.,100)
+    X,Y = np.meshgrid(x,y)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_streamplot.py]: Add space after commas in above 3 lines.

File: lib/matplotlib/tests/test_streamplot.py [hunk 13]
```diff
@@ -57,6 +65,23 @@ def test_masks_and_nans():
         plt.streamplot(X, Y, U, V, color=U, cmap=plt.cm.Blues)
 
 
+@image_comparison(baseline_images=['streamplot_maxlength'],
+                  extensions=['png'])
+def test_maxlength():
+    x, y, U, V = swirl_velocity_field()
+    plt.streamplot(x, y, U, V, maxlength=10., start_points=[[0.,1.5]],
+                   linewidth=2,density=2)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_streamplot.py]: Add space after commas in above two lines.

File: lib/matplotlib/tests/test_streamplot.py [hunk 14]
```diff
@@ -57,6 +65,23 @@ def test_masks_and_nans():
...
+
+
+@image_comparison(baseline_images=['streamplot_direction'],
+                  extensions=['png'])
+def test_direction():
+    x, y, U, V = swirl_velocity_field()
+    plt.streamplot(x, y, U, V, integration_direction='backward',
+                   maxlength=1.5, start_points=[[1.5,0.]],
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_streamplot.py]: Space after comma.

File: lib/matplotlib/pyplot.py (hunk 7 above)
Comment (gzahl) [inline lib/matplotlib/pyplot.py]: Yes, but this is generated by boilerplate.py. Maybe it was changed in another commit, but boilerplate.py was not run? I rerun boilerplate.py in the 9634c0e.

**Comment by gzahl** [comment]
I have done the requested revisions.
I don't know what to do about the unrelated change in pyplot.py. It was autogenerated by boilerplate.py (see comment above). Thanks!

File: lib/matplotlib/pyplot.py (hunk 7 above)
Comment (Kojoley) [inline lib/matplotlib/pyplot.py]: I think it is because `pyplot` was not updated with f1b89b209e9898e7f87ba3cb4f41b62aab6c82ed change

File: lib/matplotlib/streamplot.py [hunk 15]
```diff
@@ -95,6 +100,15 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
     line_kw = {}
     arrow_kw = dict(arrowstyle=arrowstyle, mutation_scale=10 * arrowsize)
 
+    if integration_direction not in ['both', 'forward', 'backward']:
+        errstr = ("Integration direction '%s' not recognised."
```
Comment (Kojoley) [inline lib/matplotlib/streamplot.py]: You need a space after the dot.

Comment (gzahl) [inline lib/matplotlib/streamplot.py]: Thanks, i fixed it (only line 104, right?)

File: lib/matplotlib/pyplot.py (hunk 7 above)
Comment (gzahl) [inline lib/matplotlib/pyplot.py]: That does seem right. Should this be fixed first, in another issue?

Comment (tacaswell) [inline lib/matplotlib/pyplot.py]: It's fine to update this here.  We should be better about not letting pyplot get 'stale', but it is what it is.

Comment (gzahl) [inline lib/matplotlib/pyplot.py]: Okay. I guess this request can therefore be closed.
I rebase again yesterday. Again there are unrelated changes from boilerplate.py (4775d7e
).

**Comment by gzahl** [comment]
@QuLogic Github still shows me a requested change, but i think i did all revisions? Otherwise please leave me a hint what is missing. Thanks!

File: lib/matplotlib/streamplot.py [hunk 16]
```diff
@@ -58,6 +59,10 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
         In data coordinates, the same as the ``x`` and ``y`` arrays.
     *zorder* : int
         any number
+    *maxlength* : float
```
Comment (NelleV) [inline lib/matplotlib/streamplot.py]: Can you please remove the "*". They are not helpful for readability.

File: lib/matplotlib/streamplot.py [hunk 17]
```diff
@@ -58,6 +59,10 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
...
+        Maximum length of streamline in axes coordinates.
+    *integration_direction* : ['forward', 'backward', 'both']
```
Comment (NelleV) [inline lib/matplotlib/streamplot.py]: Can you specify the default in the docstring?

File: lib/matplotlib/streamplot.py [hunk 18]
```diff
@@ -401,7 +416,7 @@ class TerminateTrajectory(Exception):
 # Integrator definitions
 #========================
 
-def get_integrator(u, v, dmap, minlength):
+def get_integrator(u, v, dmap, minlength, maxlength, integration_direction):
```
Comment (NelleV) [inline lib/matplotlib/streamplot.py]: I don't think this needs to be public. Can you make this function private?

**Comment by NelleV** [comment]
Hi @gzahl 
It looks good overall, but do you mind making that new function private? Else, it adds this to the list of elements that need to be backward compatible :)
The "*" removal is just bonus points: we can merge this PR without that nitpick being fixed.

**Comment by QuLogic** [comment]
Unfortunately, that function is not new (WRT this PR, at least.)

File: lib/matplotlib/streamplot.py (hunk 17 above)
Comment (QuLogic) [inline lib/matplotlib/streamplot.py]: Oh, according to numpydoc, it should be braces `{}` instead of brackets `[]`, then the first value is the default.

**Comment by NelleV** [comment]
Indeed, it is not… (I am clearly not caffeinated enough).

File: lib/matplotlib/streamplot.py (hunk 16 above)
Comment (gzahl) [inline lib/matplotlib/streamplot.py]: Of course i could, but all parameters in this function use the "*". It seems inconsistent if i only remove them for the new "maxlength" parameter?

File: lib/matplotlib/streamplot.py (hunk 18 above)
Comment (gzahl) [inline lib/matplotlib/streamplot.py]: Ok, if it was public, i should leave it like that?

**Comment by tacaswell** [comment]
Glad this finally got merged!

@gzahl Thanks for your contribution and sorry the review took so long!

**Comment by gzahl** [comment]
Thanks @tacaswell and all the others for your patience!
